File: Binary Search/Allocate_Books.cpp
// https://www.codingninjas.com/codestudio/problems/allocate-books_1090540?leftPanelTab=0

#include <bits/stdc++.h>
using namespace std;

bool isPossibleSolution(vector<int> arr, int n, int m, int mid)
{
    int studentCount = 1;
    int pageSum = 0;

    for (int i = 0; i < n; i++)
    {
        if (arr[i] + pageSum <= mid)
        {
            pageSum += arr[i];
        }
        else
        {
            studentCount++;
            if (studentCount > m || arr[i] > mid)
            {
                return false;
            }
            pageSum = arr[i];
        }
    }

    return true;
}

int allocateBooks(vector<int> arr, int n, int m)
{
    int start = 0, end = accumulate(arr.begin(), arr.end(), 0);
    int mid = start + (end - start) / 2;

    int ans = -1;

    while (start <= end)
    {
        if (isPossibleSolution(arr, n, m, mid))
        {
            ans = mid;
            end = mid - 1;
        }
        else
        {
            start = mid + 1;
        }
    }

    return ans;
}

int main()
{

    return 0;
}

File: Binary Search/Allocate_Books.py
from os import *
from sys import *
from collections import *
from math import *


def isPossible(arr, n, m, mid):
    studentCount = 1
    pageCount = 0

    for i in range(n):
        if (pageCount + arr[i]) <= mid:
            pageCount += arr[i]
        else:
            studentCount += 1
            if (studentCount > m) or (arr[i] > mid):
                return False

            pageCount = arr[i]

    return True


def allocateBooks(arr, n, m):
    start, end = 0, sum(arr)
    ans = -1
    mid = start + (end - start) // 2

    while start <= end:
        if isPossible(arr, n, m, mid) == True:
            ans = mid
            end = mid - 1
        else:
            start = mid + 1

        mid = start + (end - start) // 2

    return ans


# & Solution try
print(allocateBooks([12, 34, 67, 90], 4, 2))

"""
* The problem can easily be solved by defining the search space.
* Defining the search space for binary searching:
    ^ lower bound of the search space is the single book with minimum pages
    ^ upper bound of the search space isn't specific but it is combine inside the sum of the entire list or array
"""
